### ISSUE TITLE
Universal VM - Exploring the design space

### ISSUE THREAD

**Issue by void4** [void4/notes#14]
Described here https://github.com/void4/notes/issues/5
**tl;dr: A VM for running untrusted code**

## Syntax
Determinism is required in order to compare runtimes and their results across different machines. The runtime format (code+data+state) should therefore be deterministic. Each runtime -> runtime' transition has to be unique. The execution "worldline" therefore represents the repeated evaluation of the same (partial) function. 

#### Access control
Capabilities
Revoking access

###### Runtime and memory
Control policies are easier to implement if the system under consideration is predictable. But a deterministic system is predictable only to some extent. Instruction-level determinism may still leave the runtime unspecified or highly inconsistent: consider for example a multiplication operation with two variable sized operands (e.g. BigInts). A (static) type system might alleviate this, but runtime limitation systems based on instruction count can only allow adequate timing control if execution intervals are small.

A language based on timing interrupts might not be able to guarantee liveness if operations are too time intensive and thus have to consider externalities such as hardware as well.

It might be necessary to reconcile these realities with the formal theories of process calculi and temporal logics to allow for a better analysis of runtime properties.

###### Recursive self hosted child processes
Detaching

### Remote resources
Can remote resources be internal?

### Outsourcing tasks
Contract/decision/optimisation theory might shed some light on the specific circumstances that make outsourcing tasks to external agents sensible.

#### Ensuring confidentiality
Homomorphic computation is not yet efficient enough.

#### Verifying results
Syntactic/Structural verification
Semantic verification

Locus of execution:
Own machines
Social trust networks 
- comparing results across orthogonal or reversely incentivized agents
- security-deposits with trusted third party verifiers
